hsm/hsm.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	// For now, just assume that we can use slot 0.  Ideally, we
+	// would either let the slot id be passed in, or use a
+	// specific name.
+	session, err := api.OpenSession(slots[0], pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		cleanup()
+		return nil, ErrSessionFailure
+	}
+
+	// Attempt to log into this slot using the provided PIN.
+	if err = api.Login(session, pkcs11.CKU_USER, pin); err != nil {
+		cleanup()
+		return nil, ErrLoginFailure
+	}
+
+	return &HSM{
+		api:     api,
+		session: session,
+	}, nil
+}
+
+// Close frees up the resources used by the pkcs 11 session.
+func (h *HSM) Close() error {
+	err := h.api.Finalize()
+
+	// Free the resources associated, regardless of whether the
+	// finalize was successful or not.
+	h.api.Destroy()
+
+	return err
+}
+
+func (h *HSM) DisplayInfo() error {
+	info, err := h.api.GetInfo()
+
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	fmt.Printf("HSM:\n")
+	fmt.Printf("  Name: %s\n", info.ManufacturerID)
+	fmt.Printf("  Cryptoki: %v.%v\n", info.CryptokiVersion.Major, info.CryptokiVersion.Minor)
+	fmt.Printf("  Library: %v.%v\n", info.LibraryVersion.Major, info.LibraryVersion.Minor)
+
+	return nil
+}
+
+func (h *HSM) DisplaySlotInfo() error {
+	sinfo, err := h.api.GetSessionInfo(h.session)
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	fmt.Printf("Slot 0:\n")
+	fmt.Printf("  ID: 0x%x\n", sinfo.SlotID)
+
+	tinfo, err := h.api.GetTokenInfo(sinfo.SlotID)
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	fmt.Printf("  Token label: %s\n", tinfo.Label)
+	fmt.Printf("  Serial: %s\n", tinfo.SerialNumber)
+
+	return nil
+}
+
+func (h *HSM) displayObject(o pkcs11.ObjectHandle) error {
+	attr, err := h.api.GetAttributeValue(h.session, o, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil), // CKK_RSA, CKK_EC, etc.
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, nil),    // CKO_PUBLIC_KEY, CKO_PRIVATE_KEY, etc.
+	})
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	// Render the matching attributes
+	for _, a := range attr {
+		fmt.Printf("    ")
+		switch a.Type {
+		case pkcs11.CKA_KEY_TYPE:
+			fmt.Printf("Key Type: ")
+			switch a.Value[0] {
+			case pkcs11.CKK_RSA:
+				fmt.Printf("RSA\n")
+			case pkcs11.CKK_EC:
+				fmt.Printf("EC\n")
+			default:
+				fmt.Printf("%d\n", a.Value[0])
+			}
+		case pkcs11.CKA_LABEL:
+			fmt.Printf("Label: %s\n", a.Value)
+		case pkcs11.CKA_CLASS:
+			fmt.Printf("Class: ")
+			switch a.Value[0] {
+			case pkcs11.CKO_PUBLIC_KEY:
+				fmt.Printf("Public Key\n")
+			case pkcs11.CKO_PRIVATE_KEY:
+				fmt.Printf("Private Key\n")
+			default:
+				fmt.Printf("%d\n", a.Value[0])
+			}
+		case pkcs11.CKA_ID:
+			fmt.Printf("ID: %x\n", a.Value)
+		default:
+			fmt.Printf("0x%04X: %s (len %d)\n", a.Type, hex.EncodeToString(a.Value), len(a.Value))
+		}
+	}
+
+	return nil
+}
+
+func (h *HSM) FindObjectsByLabel(label string, max int) error {
```

Comment:
We should probably return the object handles as well ... this is currently more of a debug function with a misleading name, since it just prints them for the moment.